### PR TITLE
LSTM learning rate added.

### PR DESCRIPTION
**Pull request by gtoderici**

### CODE REVIEW

**Comment by cxz** [comment]
Hi @gtoderici, have you tested this model with the suggested learning rate 0.001 and all other default parameters? For me the GAP metric does not go above 0.00.

**Comment by gtoderici** [comment]
Hi ccz, how long are you letting the code train? With 2 layers, 1024 cells, and  LR of 0.001 video model set to MoeModel it should jump up in a few thousand steps. I haven't tried the exact version of the code which is checked in right now because there were more commits after I tested it (I may give it a try later today).

For debugging, please the full command line you're using to train, and maybe I can help.

**Comment by cxz** [comment]
I used the following command

```
gcloud --verbosity=warning beta ml jobs submit training lstm_20170307_1035_train \
        --package-path=youtube-8m --module-name=youtube-8m.train \
        --staging-bucket='gs://2017-kgl' --region=us-east1 \
        --config=config/cloudml-gpu.yaml \
        -- --train_data_pattern='gs://youtube8m-ml-us-east1/1/frame_level/train/train*.tfrecord' \
		--train_dir='gs://2017-kgl/y8tm/lstm_1' \
		--frame_features=True \
		--model='LstmModel' \
		--batch_size=128 \
		--base_learning_rate=0.001 \
		--feature_names='rgb, audio' \
		--feature_sizes='1024, 128'
```

and trained on "standard_gpu" for 1 epoch (~6 hours). I kept the default of 2 layers, 1024 cells and only adjusted the batch_size to 128 and learning_rate to 0.001 as written in the documentation.

**Comment by gtoderici** [comment]
Something is seriously wrong. I am getting negative loss:
`INFO:tensorflow:/job:master/task:0: training step 23| Hit@1: 0.00 PERR: 0.00 GAP: 0.00 Loss: -0.0472238
`
I will investigate further because this should not happen.
As an interim fix, in frame_level_models, for the LstmModel class, there's a return statement which create a video level model. I am suspecting the bug could be related to that.

Please try replacing it with:
  ```
  output = slim.fully_connected(state, vocab_size,                                                                                                     
                                  activation_fn=tf.nn.sigmoid,                                                                                           
                                  weights_regularizer=slim.l2_regularizer(0.01),                                                                         
                                  scope='final_logistic')                                                                                                 
    return {"predictions": output }     
```

Can you please also paste a few lines with the loss you're observing in addition to GAP?

**Comment by cxz** [comment]
Here it is:
```
I  /job:master/task:0: training step 4239| Hit@1: 0.07 PERR: 0.02 GAP: 0.00 Loss: 93.5854 
I  /job:master/task:0: training step 4240| Hit@1: 0.02 PERR: 0.03 GAP: 0.00 Loss: 94.8446 
I  /job:master/task:0: training step 4241| Hit@1: 0.12 PERR: 0.03 GAP: 0.00 Loss: 95.924 
I  /job:master/task:0: training step 4242| Hit@1: 0.02 PERR: 0.01 GAP: 0.00 Loss: 94.215 
I  /job:master/task:0: training step 4243| Hit@1: 0.08 PERR: 0.02 GAP: 0.00 Loss: 92.4161 
```

Tensorboard:
lstm1

**Comment by LongLong-Jing** [comment]
Hi @cxz ,

Thanks for your work. Can you tell me the memory of your GPU?  I tested LSTM with 1024 cell in TITAN GPU. But I got error of our of memory even when batch size = 1. Thanks!

**Comment by gtoderici** [comment]
I am not sure what's happening, still. It seems that it's possible to get the latest code and still have a reasonable behavior. See below (latest code):

`I  /job:master/task:0: training step 1387| Hit@1: 0.55 PERR: 0.38 GAP: 0.43 Loss: 13.3166 
`
Here are the flags for my job so that my run can be reproduced. 

```
{
  "scaleTier": "CUSTOM",
  "masterType": "standard_gpu",
  "workerType": "standard",
  "parameterServerType": "standard",
  "packageUris": [
    "gs://cloud-ml-e2e-testing-us-central1/tmp/youtube8m_frame_rgb_audio_lstm_1sgpu_benchmark_run_20170308_154235/trainer.tar"
  ],
  "pythonModule": "trainer.train",
  "args": [
    "--train_dir=gs://cloud-ml-e2e-testing-us-central1/tmp/youtube8m_frame_rgb_audio_lstm_1sgpu_benchmark_run_20170308_154235/output/train",
    "--max_steps=30300",
    "--start_new_model",
    "--frame_features=True",
    "--model=LstmModel",
    "--feature_names=rgb, audio",
    "--feature_sizes=1024, 128",
    "--batch_size=128",
    "--train_data_pattern=gs://youtube8m-ml-us-east1/1/frame_level/train/train*.tfrecord",
    "--base_learning_rate=0.001"
  ],
  "region": "us-east1",
  "runtimeVersion": "1.0"
}
```

**Comment by cxz** [comment]
Replacing the Moe at the end with a fully connected layers yields:

```
23:54:44.830 /job:master/task:0: training step 540| Hit@1: 0.64 PERR: 0.41 GAP: 0.427157 Loss: 13.722
```

**Comment by gtoderici** [comment]
@LongLong-Jing - everything should run just fine on a Titan X. We're running this on cloud, on a half of a K80 (so 12 GB).

@cxz - Interesting. I think Moe should also work, though (haven't tested recently, but it did work after LSTM before).
